src/component.jl

```diff
@@ -184,29 +195,120 @@ Base.sortperm(c::SharedComponent) = sortperm(c.data)
     end
     return h
 end
+
+#TODO: These boundchecks etc are to ensure in the case of || in @entities_in nobody is trying to load some
+#      invalid memory. Do we really need ||?
+@generated function Base.getindex(t::Tup, ::Type{T}) where {Tup <: Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, Tup.parameters)
+    # return :(unsafe_load(@inbounds getindex(t, $id))::T) # without ||
+    return quote
+        $(Expr(:meta, :inline))
+        $(Expr(:meta, :propagate_inbounds))
+        ptr = getindex(t, $id)
+        @boundscheck if ptr === Ptr{T}()
+            throw(BoundsError(t, T))
+        end
+        unsafe_load(ptr)::T
+    end
+end
+
+@generated function Base.setindex!(t::Tup, x::T, ::Type{T}) where {Tup <: Tuple, T<:ComponentData}
+    id = findfirst(x -> eltype(x) == T, Tup.parameters)
+    # return :(unsafe_store!(@inbounds getindex(t, $id), x)) # without ||
+    return quote
+        $(Expr(:meta, :inline))
+        $(Expr(:meta, :propagate_inbounds))
+        ptr = getindex(t, $id)
+        @boundscheck if ptr === Ptr{T}()
+            throw(BoundsError(t, T))
+        end
+        unsafe_store!(ptr, x)
+    end
+end
 ########################################
 #                                      #
 #            Iteration                 #
 #                                      #
 ########################################
-struct EntityIterator{T <: Union{IndicesIterator,Indices,AbstractGroup}}
+struct EntityIterator{T <: Union{IndicesIterator,Indices,AbstractGroup}, TT <: Tuple}
     it::T
+    components::TT
 end
 
-Base.eltype(::EntityIterator) = Entity
+Base.eltype(::EntityIterator{T, TT}) where {T, TT} = EntityState{Tuple{map(x->Ptr{eltype(x)}, TT.parameters)...}}
```

Comment:
With the `map` over `TT.parameters` I guess this won't be type-inferred? (Maybe not a problem but could be worth checking)

Reply:
I was too hopeful the compiler would do magic

Reply:
Actually, using `@generated` also doesn't seem to do the trick, how do I do this to be type-inferred?

Reply:
Yeah that's kind of surprising, but `@generated` should also be overkill anyway. Marking this as `Base.@pure` is probably the easiest and "correct enough" in this specific case.

Failing that, you might be able do something tricky with recursive functions and tuples... but it's quite tricky and I don't immediately see how because computing with types rather than values is quite awkward.

Reply:
seems @pure also doesn't do the trick :disappointed:

Reply:
Ah, It's a subtle bug in the implementation: you splatted the tuple of types in at *runtime*, but it should have been splatted in at expression creation time

`$(t...)`  vs  `$t...`